WIP Parameter names are normalized to not contain colons

mrclay opened the pull request:
Fetched array keys do not contain colons.
For BC, params/types can be still be accessed via names with colons.
Invalid names result in QueryException.

mrclay commented on tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php:
```diff
@@ -882,4 +882,55 @@ public function testJoinWithNonUniqueAliasThrowsException()
+
+        $this->assertSame(array(
+            'foo' => 'foo',
+            'bar' => 2,
```
The big question is whether this is BC. If user code _expects_ a `:bar` key, it will fail. execute() handles it of course.

Ocramius replied:
Getting parameters is a rare use-case, but we can document the breakage here. This one seems to be the only affected endpoint

mrclay replied:
FYI I'm already normalizing in my subclass so this isn't a high priority for me. If you'd rather me just update the docs accordingly to avoid BC changes I can do that instead.

Ocramius replied:
> If you'd rather me just update the docs accordingly to avoid BC changes I can do that instead.

Updating [`UPGRADE.md`](https://github.com/doctrine/dbal/blob/61e1f860a304c3fb6961ecbec36927b5f20b71c6/UPGRADE.md) is sufficient

mrclay commented on tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php:
```diff
@@ -882,4 +882,55 @@ public function testJoinWithNonUniqueAliasThrowsException()
+            "The given name '1a' is invalid."
+        );
+
+        $qb->setParameter('1a', 'one a');
```
The plan here is to use PDO's validation strategy http://stackoverflow.com/a/5810058/3779 but allow string integers like "1" to pass through.

Ocramius commented on .gitignore:
```diff
@@ -4,5 +4,6 @@ reports/
 dist/
 download/
 vendor/
+.idea/
```
This should be in your global `.gitignore` instead

Ocramius commented on lib/Doctrine/DBAL/Query/QueryBuilder.php:
```diff
@@ -1329,6 +1348,36 @@ private function getSQLForJoins($fromAlias, array &$knownAliases)
+        }
+
+        // int as string
+        if (preg_match('~^(0|[1-9]\\d*)\\z~', $key)) {
```
Can be simplified to `$key === (string) (int) $key`

Ocramius commented on lib/Doctrine/DBAL/Query/QueryException.php:
```diff
@@ -51,4 +51,14 @@ static public function nonUniqueAlias($alias, $registeredAliases)
+     *
+     * @return \Doctrine\DBAL\Query\QueryException
+     */
+    static public function invalidParamName($name)
```
don't abbreviate `Parameter`

Ocramius commented on lib/Doctrine/DBAL/Query/QueryException.php:
```diff
@@ -51,4 +51,14 @@ static public function nonUniqueAlias($alias, $registeredAliases)
+    /**
+     * @param string $name
+     *
+     * @return \Doctrine\DBAL\Query\QueryException
```
`@return self` is also sufficient

mrclay replied:
Mind if I send another PR to change all the fluent methods to use `self`?

Ocramius replied:
Not worth the effort, imo, but appreciated :-)

Ocramius commented on tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php:
```diff
@@ -882,4 +882,77 @@ public function testJoinWithNonUniqueAliasThrowsException()
+        $this->assertSame(2, $qb->getParameter('bar'));
+        $this->assertSame(2, $qb->getParameter(':bar'));
+
+        $this->assertSame(array(
```
You can use short arrays: we dropped 5.3 support ages ago :-)

Ocramius commented on tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php:
```diff
@@ -882,4 +882,77 @@ public function testJoinWithNonUniqueAliasThrowsException()
+        $this->assertSame(array(
+            'foo' => 'foo',
+            'bar' => 2,
+        ), $qb->getParameters());
```
Note that this formatting is not PSR-2 compliant

mrclay committed:
Parameter names are normalized to not contain colons

Fetched array keys do not contain colons.
For BC, params/types can be still be accessed via names with colons.
Invalid names result in QueryException.

mrclay commented on UPGRADE.md:
```diff
@@ -1,3 +1,15 @@
+## MINOR BC BREAK: Doctrine\DBAL\Query\QueryBuilder
```
What version do you think this will be in? I figured I'd just leave this out for now.

Ocramius replied:
@mrclay 2.6.0

mrclay commented:
Technically PDO should accept parameters not only starting with digits, but containing only digits, like `:0` or `:00`. I kinda don't know what to do with this fact. Such keys would probably cause problems in DBAL's execution strategy, but a key like `1a` shouldn't cause a problem (and might be used with this class today).

Strawman: If given a key containing only digits, assume it's positional (like this PR). If that key is prepended with `:` or if PHP will interpret the string as an integer key (e.g. `1e2`), throw.

Positional keys: `1`, `"1"`
Named keys: `"foo"`, `"9lives"`
Rejected: `":1"`, `"1e9"`
